akka-http-tests/src/test/java/akka/http/javadsl/server/directives/HeaderDirectivesTest.java

```diff
@@ -0,0 +1,208 @@
+      (Server s) -> complete(s.getProducts().iterator().next().product())));
+
+    route
+      .run(HttpRequest.create().addHeader(Server.create(ProductVersion.create("such-service", "0.6"))))
+      .assertStatusCode(StatusCodes.OK)
+      .assertEntity("such-service");
+
+    route
+      .run(HttpRequest.create())
+      .assertStatusCode(StatusCodes.BAD_REQUEST);
+  }
+
+  @Test
+  public void testOptionalHeaderValue() {
+    TestRoute route = testRoute(optionalHeaderValue((header) -> {
+        if (header.name().equals("X-Test-Header")) {
+          if (header.value().equals("bad value")) throw new RuntimeException("bad value");
+          else return Optional.of(header.value());
+        }
+        else return Optional.empty();
+      },
+      opt -> complete(opt.toString())));
+
+    route
+      .run(HttpRequest.create().addHeader(RawHeader.create("X-Test-Header", "woho!")))
+      .assertStatusCode(StatusCodes.OK)
+      .assertEntity("Optional[woho!]");
+
+    route
+      .run(HttpRequest.create().addHeader(RawHeader.create("X-Test-Header", "bad value")))
+      .assertStatusCode(StatusCodes.BAD_REQUEST);
+
+
+    route
+      .run(HttpRequest.create())
+      .assertStatusCode(StatusCodes.OK)
+      .assertEntity("Optional.empty");
+  }
+
+  @Test
+  public void testOptionalHeaderValuePF() {
+    TestRoute route = testRoute(optionalHeaderValuePF(
+      new PFBuilder<HttpHeader, String>().<Host>match(
+        Host.class, Host::value
+      ).build(),
+      (opt) -> complete(opt.toString())));
+
+    route
+      .run(HttpRequest.create().addHeader(Host.create("example.com")))
+      .assertStatusCode(StatusCodes.OK)
+      .assertEntity("Optional[example.com]");
+
+    route
+      .run(HttpRequest.create())
+      .assertStatusCode(StatusCodes.OK)
+      .assertEntity("Optional.empty");
+  }
+
+
+  @Test
+  public void testOptionalHeaderValueByName() {
+    TestRoute route = testRoute(optionalHeaderValueByName("X-Test-Header", (opt) -> complete(opt.toString())));
+
+    route
+      .run(HttpRequest.create().addHeader(RawHeader.create("X-Test-Header", "woho!")))
+      .assertStatusCode(StatusCodes.OK)
+      .assertEntity("Optional[woho!]");
+
+    route
+      .run(HttpRequest.create())
+      .assertStatusCode(StatusCodes.OK)
+      .assertEntity("Optional.empty");
+  }
+
+  @Test
+  public void testOptionalHeaderValueByType() {
+    TestRoute route = testRoute(optionalHeaderValueByType(Server.class,
+      (Optional<Server> s) -> complete(((Boolean)s.isPresent()).toString())));
+
+    route
+      .run(HttpRequest.create().addHeader(Server.create(ProductVersion.create("such-service", "0.6"))))
+      .assertStatusCode(StatusCodes.OK)
+      .assertEntity("true");
+
+    route
+      .run(HttpRequest.create())
+      .assertStatusCode(StatusCodes.OK)
+      .assertEntity("false");
+  }
+
+
+  public static class MyHeader extends ModeledCustomHeader {
+
+    private final String value;
+
+    public MyHeader(String value) {
```

Comment:
This is accidental though I think.
Make the MyHeader "weird", as in: `public MyHeader(Long shoeSize, String name) {` and hit it with a request like: `X-My-Header: 12-person`